pageserver/src/storage_sync.rs

```diff
@@ -129,1330 +114,587 @@
+        .await
+    }
 
-/// Adds the new files to delete as a deletion task to the queue.
-/// On task failure, it gets retried again from the start a number of times.
-///
-/// Ensure that the loop is started otherwise the task is never processed.
-pub fn schedule_layer_delete(
-    tenant_id: TenantId,
-    timeline_id: TimelineId,
-    layers_to_delete: HashSet<PathBuf>,
-) {
-    let sync_queue = match SYNC_QUEUE.get() {
-        Some(queue) => queue,
-        None => {
-            warn!("Could not send deletion task for tenant {tenant_id}, timeline {timeline_id}");
-            return;
-        }
-    };
-    sync_queue.push(
-        TenantTimelineId {
-            tenant_id,
-            timeline_id,
-        },
-        SyncTask::delete(LayersDeletion {
-            layers_to_delete,
-            deleted_layers: HashSet::new(),
-            deletion_registered: false,
-        }),
-    );
-    debug!("Deletion task for tenant {tenant_id}, timeline {timeline_id} sent")
-}
+    //
+    // Upload operations.
+    //
 
-/// Requests the download of the entire timeline for a given tenant.
-/// No existing local files are currently overwritten, except the metadata file (if its disk_consistent_lsn is less than the downloaded one).
-/// The metadata file is always updated last, to avoid inconsistencies.
-///
-/// On any failure, the task gets retried, omitting already downloaded layers.
-///
-/// Ensure that the loop is started otherwise the task is never processed.
-pub fn schedule_layer_download(tenant_id: TenantId, timeline_id: TimelineId) {
-    debug!("Scheduling layer download for tenant {tenant_id}, timeline {timeline_id}");
-    let sync_queue = match SYNC_QUEUE.get() {
-        Some(queue) => queue,
-        None => {
-            warn!("Could not send download task for tenant {tenant_id}, timeline {timeline_id}");
-            return;
-        }
-    };
-    sync_queue.push(
-        TenantTimelineId {
-            tenant_id,
-            timeline_id,
-        },
-        SyncTask::download(LayersDownload {
-            layers_to_skip: HashSet::new(),
-        }),
-    );
-    debug!("Download task for tenant {tenant_id}, timeline {timeline_id} sent")
-}
+    ///
+    /// Launch an index-file upload operation in the background.
+    ///
+    /// The upload will be added to the queue immediately, but it
+    /// won't be performed until all previosuly scheduled layer file
+    /// upload operations have completed successfully.  This is to
+    /// ensure that when the index file claims that layers X, Y and Z
+    /// exist in remote storage, they really do.
+    pub fn schedule_index_upload(
+        self: &Arc<Self>,
+        metadata: Option<&TimelineMetadata>,
+    ) -> Result<()> {
+        // TODO: Construct IndexPart, attach the metadata to it
+
+        let mut upload_queue = self.upload_queue.lock().unwrap();
+        ensure!(
+            upload_queue.latest_metadata.is_some(),
+            "upload queue not initialized"
```

Comment:
The pattern occurs 4 times in the code and I wonder, if we could make some one-step `new` to init the queue?
We could pause any sync operations before metadata and other data is set for the first time: this way, we would keep init state checks in one place.